In [1]:
import trainer.glue_base as glue_base
import models.sparse_token as sparse
import pickle, importlib
importlib.reload(glue_base)
importlib.reload(sparse)
Glue = glue_base.GlueAttentionApproxTrainer

In [2]:
subsets = ["mnli","mrpc","qnli","qqp","rte","sst2","stsb"]
subsets = ["cola","mnli","mrpc"]#,"qnli","qqp","rte","sst2","stsb","wnli",]
kss = [
    0.1, 0.25, 0.375, 0.5, 0.625, 0.75, 0.875, 0.999, 'dynamic', 
    # 'dynamic:avg:avg:true', 'dynamic:avg:avg:false', 'dynamic:avg:max:true', 'dynamic:avg:max:false',
    # 'dynamic:max:avg:true', 'dynamic:max:avg:false', 'dynamic:max:max:true', 'dynamic:max:max:false',
]
sparse.benchmark_reset()
# subsets = ["mrpc"]
# kss = ['dynamic:avg:avg:f',0.1]

def get_score(score):
    if 'accuracy' in score:
        return score['accuracy'], "acc"
    first_metric = list(score.keys())[0]
    return score[first_metric], first_metric

results = {}
i = 0
for subset in subsets:
    trainer = Glue(dataset=subset, factor=16, batch_size=-1, device=0)
    trainer.load()
    scores = {}
    metric_name = ""
    bert_score, metric_name = get_score(trainer.eval_base_model())
    scores['bert'] = f'{bert_score:.5f}'
    for ks in kss:
        sparse.benchmark_reset()
        sparse_score, _ = get_score(trainer.eval_sparse_model(ks=ks))
        if isinstance(ks, str) and ks.startswith('dynamic'):
            est_k = sparse.benchmark_get_average('est_k')
            scores[str(ks)] = f'{sparse_score:.5f} (k:{est_k:.2f})'
        else:
            scores[str(ks)] = f'{sparse_score:.5f}'
        i += 1
        count = len(subsets) * len(kss)
        print(f'{i}/{count} | {subset} {ks} = {sparse_score:.5f}')
    results[f"{subset} ({metric_name})"] = scores

with open('glue_benchmark.pkl', 'wb') as f:
    pickle.dump(results, f)

sparse.benchmark_report()

Trainer: cola


Reusing dataset glue (./cache/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad)
Loading cached processed dataset at ./cache/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-6b1dfe928518b6a9.arrow
Loading cached processed dataset at ./cache/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-f7e44aa0ee690f24.arrow
Reusing dataset glue (./cache/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad)
Loading cached processed dataset at ./cache/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-99bc25cbc8b7096f.arrow
Loading cached processed dataset at ./cache/datasets/glue/cola/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-49888407280455f9.arrow


Trainer: Save checkpoint path saves/glue-cola-16.pth


eval: 100%|██████████| 131/131 [00:05<00:00, 24.69it/s]


metric score {'matthews_correlation': 0.5338774230813111}
avg occupy 0.34522852946633237


eval: 100%|██████████| 131/131 [00:10<00:00, 12.35it/s]


metric score {'matthews_correlation': 0.05521043543404519}
avg occupy 0.34522852946633237
1/27 | cola 0.1 = 0.05521


eval: 100%|██████████| 131/131 [00:11<00:00, 11.31it/s]


metric score {'matthews_correlation': 0.408526513461678}
avg occupy 0.34522852946633237
2/27 | cola 0.25 = 0.40853


eval: 100%|██████████| 131/131 [00:10<00:00, 12.45it/s]


metric score {'matthews_correlation': 0.4586369195095741}
avg occupy 0.34522852946633237
3/27 | cola 0.375 = 0.45864


eval: 100%|██████████| 131/131 [00:08<00:00, 15.53it/s]


metric score {'matthews_correlation': 0.5025517897100551}
avg occupy 0.34522852946633237
4/27 | cola 0.5 = 0.50255


eval: 100%|██████████| 131/131 [00:07<00:00, 17.03it/s]


metric score {'matthews_correlation': 0.5259790835948196}
avg occupy 0.34522852946633237
5/27 | cola 0.625 = 0.52598


eval: 100%|██████████| 131/131 [00:07<00:00, 16.55it/s]


metric score {'matthews_correlation': 0.5311686988596115}
avg occupy 0.34522852946633237
6/27 | cola 0.75 = 0.53117


eval: 100%|██████████| 131/131 [00:07<00:00, 17.20it/s]


metric score {'matthews_correlation': 0.5311983410233877}
avg occupy 0.34522852946633237
7/27 | cola 0.875 = 0.53120


eval: 100%|██████████| 131/131 [00:08<00:00, 16.20it/s]


metric score {'matthews_correlation': 0.5338774230813111}
avg occupy 0.34522852946633237
8/27 | cola 0.999 = 0.53388


eval: 100%|██████████| 131/131 [00:07<00:00, 16.93it/s]


metric score {'matthews_correlation': 0.5338774230813111}
avg occupy 0.34522852946633237
9/27 | cola dynamic = 0.53388
Trainer: mnli


Reusing dataset glue (./cache/datasets/glue/mnli/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad)
Loading cached processed dataset at ./cache/datasets/glue/mnli/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-c8de9ec562ad8c1a.arrow
Loading cached processed dataset at ./cache/datasets/glue/mnli/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-e8a73fc1e1933d2b.arrow
Reusing dataset glue (./cache/datasets/glue/mnli/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad)
Loading cached processed dataset at ./cache/datasets/glue/mnli/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-baccd56b3c558104.arrow


  0%|          | 0/10 [00:00<?, ?ba/s]

Trainer: Save checkpoint path saves/glue-mnli-16.pth


eval: 100%|██████████| 1227/1227 [01:00<00:00, 20.12it/s]


metric score {'accuracy': 0.8420784513499745}
avg occupy 0.2057366697815377


eval: 100%|██████████| 1227/1227 [01:10<00:00, 17.47it/s]


metric score {'accuracy': 0.5840040753948039}
avg occupy 0.2057366697815377
10/27 | mnli 0.1 = 0.58400


eval: 100%|██████████| 1227/1227 [01:12<00:00, 16.90it/s]


metric score {'accuracy': 0.8178298522669384}
avg occupy 0.2057366697815377
11/27 | mnli 0.25 = 0.81783


eval: 100%|██████████| 1227/1227 [01:18<00:00, 15.68it/s]


metric score {'accuracy': 0.8383087111563933}
avg occupy 0.2057366697815377
12/27 | mnli 0.375 = 0.83831


eval: 100%|██████████| 1227/1227 [01:23<00:00, 14.66it/s]


metric score {'accuracy': 0.8416709118695873}
avg occupy 0.2057366697815377
13/27 | mnli 0.5 = 0.84167


eval: 100%|██████████| 1227/1227 [01:25<00:00, 14.36it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
14/27 | mnli 0.625 = 0.84198


eval: 100%|██████████| 1227/1227 [01:29<00:00, 13.63it/s]


metric score {'accuracy': 0.8419765664798777}
avg occupy 0.2057366697815377
15/27 | mnli 0.75 = 0.84198


eval: 100%|██████████| 1227/1227 [01:30<00:00, 13.62it/s]


metric score {'accuracy': 0.8420784513499745}
avg occupy 0.2057366697815377
16/27 | mnli 0.875 = 0.84208


eval: 100%|██████████| 1227/1227 [01:39<00:00, 12.31it/s]


metric score {'accuracy': 0.8420784513499745}
avg occupy 0.2057366697815377
17/27 | mnli 0.999 = 0.84208


eval: 100%|██████████| 1227/1227 [01:20<00:00, 15.27it/s]


metric score {'accuracy': 0.841874681609781}
avg occupy 0.2057366697815377
18/27 | mnli dynamic = 0.84187
Trainer: mrpc


Reusing dataset glue (./cache/datasets/glue/mrpc/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad)
Loading cached processed dataset at ./cache/datasets/glue/mrpc/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-9afadd04a309a170.arrow
Loading cached processed dataset at ./cache/datasets/glue/mrpc/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-0bddda4d866e6dbd.arrow
Reusing dataset glue (./cache/datasets/glue/mrpc/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad)
Loading cached processed dataset at ./cache/datasets/glue/mrpc/1.0.0/dacbe3125aa31d7f70367a07a8a9e72a5a0bfeb5fc42e75c9db75b96da6053ad/cache-efab46bb85d6ed58.arrow


  0%|          | 0/2 [00:00<?, ?ba/s]

Trainer: Save checkpoint path saves/glue-mrpc-16.pth


eval: 100%|██████████| 216/216 [00:07<00:00, 30.48it/s]


metric score {'accuracy': 0.8440579710144928, 'f1': 0.8865457612821593}
avg occupy 0.5399536083129833


eval: 100%|██████████| 216/216 [00:12<00:00, 17.42it/s]


metric score {'accuracy': 0.33507246376811595, 'f1': 0.0}
avg occupy 0.5399536083129833
19/27 | mrpc 0.1 = 0.33507


eval: 100%|██████████| 216/216 [00:10<00:00, 19.66it/s]


metric score {'accuracy': 0.34550724637681157, 'f1': 0.035866780529461996}
avg occupy 0.5399536083129833
20/27 | mrpc 0.25 = 0.34551


eval: 100%|██████████| 216/216 [00:12<00:00, 17.44it/s]


metric score {'accuracy': 0.4405797101449275, 'f1': 0.3170559094125973}
avg occupy 0.5399536083129833
21/27 | mrpc 0.375 = 0.44058


eval: 100%|██████████| 216/216 [00:12<00:00, 17.69it/s]


metric score {'accuracy': 0.6539130434782608, 'f1': 0.6842940243257536}
avg occupy 0.5399536083129833
22/27 | mrpc 0.5 = 0.65391


eval: 100%|██████████| 216/216 [00:12<00:00, 17.26it/s]


metric score {'accuracy': 0.807536231884058, 'f1': 0.8525754884547068}
avg occupy 0.5399536083129833
23/27 | mrpc 0.625 = 0.80754


eval: 100%|██████████| 216/216 [00:12<00:00, 17.58it/s]


metric score {'accuracy': 0.84, 'f1': 0.8828522920203735}
avg occupy 0.5399536083129833
24/27 | mrpc 0.75 = 0.84000


eval: 100%|██████████| 216/216 [00:13<00:00, 16.61it/s]


metric score {'accuracy': 0.8440579710144928, 'f1': 0.8865457612821593}
avg occupy 0.5399536083129833
25/27 | mrpc 0.875 = 0.84406


eval: 100%|██████████| 216/216 [00:11<00:00, 18.65it/s]


metric score {'accuracy': 0.8440579710144928, 'f1': 0.8865457612821593}
avg occupy 0.5399536083129833
26/27 | mrpc 0.999 = 0.84406


eval: 100%|██████████| 216/216 [00:12<00:00, 16.73it/s]


metric score {'accuracy': 0.8440579710144928, 'f1': 0.8865457612821593}
avg occupy 0.5399536083129833
27/27 | mrpc dynamic = 0.84406
est_k: (2592, 1894.1703296703308), avg: 0.7307755901505906


In [3]:
import pickle
import pandas as pd

with open('glue_benchmark.pkl', 'rb') as f:
    results = pickle.load(f)

data = []
subsets = list(results.keys())
factors = list(results[subsets[0]].keys())
for factor in factors:             
    row = []
    for subset in subsets:
        row.append(results[subset][factor])
    data.append(row)

#calculate reproducibility579
data_scalar = []
for line in data:
    xs = []
    for item in line:
        xs.append(float(item.split()[0]))
    data_scalar.append(xs)
reproducibilities = []
for i in range(len(data_scalar)):
    rsum = 0
    for k in range(len(data_scalar[i])):
        rsum += data_scalar[i][k]/data_scalar[0][k]
    rsum /= len(data_scalar[i])
    reproducibilities.append(rsum)
for i, r in enumerate(reproducibilities):
    data[i].append(f"{r*100:.2f}")
subsets.append("reproduce")

df = pd.DataFrame(data, columns=subsets, index=factors)
tex = df.to_latex()
with open('saves_plot/glue_benchmark.tex', 'w') as f:
    f.write(tex)
df

cola (matthews_correlation)        mnli (acc)        mrpc (acc)  \
bert                        0.53388           0.84208           0.84406   
0.1                         0.05521           0.58400           0.33507   
0.25                        0.40853           0.81783           0.34551   
0.375                       0.45864           0.83831           0.44058   
0.5                         0.50255           0.84167           0.65391   
0.625                       0.52598           0.84198           0.80754   
0.75                        0.53117           0.84198           0.84000   
0.875                       0.53120           0.84208           0.84406   
0.999                       0.53388           0.84208           0.84406   
dynamic            0.53388 (k:0.48)  0.84187 (k:0.37)  0.84406 (k:0.73)   

        reproduce  
bert       100.00  
0.1         39.80  
0.25        71.53  
0.375       79.22  
0.5         90.52  
0.625       98.06  
0.75        99.67  
0.875       99.83  
0.999      100.00  
dynamic     99.99

	cola (matthews_correlation)	mnli (acc)	mrpc (acc)	qnli (acc)	qqp (acc)	rte (acc)	sst2 (acc)	stsb (pearson)	wnli (acc)	reproduce
bert	0.53388	0.84198	0.84406	0.91543	0.90908	0.72563	0.92431	0.88047	0.56338	100.00
0.1	0.02214	0.57932	0.33507	0.59217	0.73391	0.52347	0.70872	0.01637	0.60563	57.36
0.25	0.33346	0.81559	0.36174	0.83892	0.89115	0.70758	0.85092	0.55826	0.56338	82.76
0.375	0.45301	0.83770	0.54377	0.89731	0.90616	0.70758	0.89335	0.75968	0.56338	91.88
0.5	0.49403	0.84167	0.76928	0.90976	0.90885	0.73285	0.90596	0.84956	0.56338	97.61
0.625	0.51112	0.84167	0.82957	0.91360	0.90915	0.72563	0.91628	0.87337	0.56338	99.12
0.75	0.52328	0.84198	0.84290	0.91433	0.90913	0.72924	0.92202	0.87936	0.56338	99.77
0.875	0.53120	0.84208	0.84348	0.91525	0.90913	0.72563	0.92317	0.88039	0.56338	99.92
0.999	0.53388	0.84198	0.84406	0.91543	0.90908	0.72563	0.92431	0.88047	0.56338	100.00

In [ ]:
# trainer = Glue(dataset='qnli', factor=16, batch_size=-1, device=0)
# trainer.load() 

In [ ]:
# trainer.model.bert = trainer.model_bert
# trainer.eval_base_model()

In [ ]:
# trainer.seed()
# import models.sparse_token as sparse
# import transformers.models.bert.modeling_bert as berts
# import importlib
# importlib.reload(sparse)

# wrapped_bert = sparse.ApproxSparseBertModel(trainer.model_bert, approx_bert=trainer.approx_bert, ks=0.1)
# sparse_cls_bert = berts.BertForSequenceClassification(trainer.model_bert.config)
# sparse_cls_bert.load_state_dict(trainer.model.state_dict())
# sparse_cls_bert.bert = wrapped_bert
# sparse_cls_bert.to(trainer.device).eval()

# trainer.eval_base_model(model = sparse_cls_bert)